Sources/Components/CarouselView/CarouselView.swift

```diff
@@ -0,0 +1,192 @@
+    func numberOfItems(in carouselView: CarouselView) -> Int
+    func carouselView(_ carouselView: CarouselView, cellForItemAt indexPath: IndexPath) -> CarouselViewCell
+}
+
+public protocol CarouselViewDelegate: AnyObject {
+    func carouselView(_ carouselView: CarouselView, didSelectItemAt indexPath: IndexPath)
+    func carouselViewDidEndDecelerating(_ carouselView: CarouselView)
+}
+
+public class CarouselView: UIView {
+
+    public weak var dataSource: CarouselViewDataSource? {
+        didSet {
+            collectionView.reloadData()
+        }
+    }
+
+    public weak var delegate: CarouselViewDelegate?
+
+    // MARK: - Private Properties
+
+    private var initialLayout = true
+
+    private lazy var layout: UICollectionViewFlowLayout = {
+        let layout = UICollectionViewFlowLayout()
+        layout.minimumLineSpacing = 0
+        layout.minimumInteritemSpacing = 0
+        layout.scrollDirection = .horizontal
+        return layout
+    }()
+
+    private lazy var collectionView: UICollectionView = {
+        let collectionView = UICollectionView(frame: .zero, collectionViewLayout: layout)
+        collectionView.delegate = self
+        collectionView.dataSource = self
+        collectionView.isPagingEnabled = true
+        collectionView.showsHorizontalScrollIndicator = false
+        collectionView.translatesAutoresizingMaskIntoConstraints = false
+        return collectionView
+    }()
+
+    // MARK: - Init
+
+    public override init(frame: CGRect) {
+        super.init(frame: frame)
+        setup()
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    // MARK: - Overrides
+
+    public override var backgroundColor: UIColor? {
+        didSet {
+            collectionView.backgroundColor = backgroundColor
+        }
+    }
+
+    public override func layoutSubviews() {
+        super.layoutSubviews()
+
+        guard initialLayout else { return }
+        initialLayout = false
+
+        // Scroll to first item
+        collectionView.contentOffset = CGPoint(
+            x: bounds.width,
+            y: collectionView.contentOffset.y
+        )
+    }
+}
+
+public extension CarouselView {
+    var contentOffset: CGPoint {
+        return CGPoint(
+            x: collectionView.contentOffset.x - collectionView.bounds.width,
+            y: collectionView.contentOffset.y
+        )
+    }
+
+    func register(_ cellClass: CarouselViewCell.Type) {
+        collectionView.register(cellClass)
+    }
+
+    func dequeue<T>(_ cellType: T.Type, for indexPath: IndexPath) -> T where T: CarouselViewCell {
+        return collectionView.dequeue(cellType, for: indexPath)
+    }
+
+    func scrollToItem(at indexPath: IndexPath, at scrollPosition: UICollectionView.ScrollPosition, animated: Bool) {
+        collectionView.scrollToItem(at: IndexPath(item: indexPath.item + 1, section: 0), at: scrollPosition, animated: animated)
```

Comment:
Since this function is public, I guess it's possible to crash the app when attempting to scroll to invalid index path. Perhaps some index path validation is needed here.

Reply:
That is probably a good idea.